### PR TITLE
Experimental docker support for running openstack projects

### PR DESCRIPTION
**Pull request by paulczar**
Very early and experimental support for running openstack projects in docker containers.

HEAT used as an example.

basic premise ...   
- docker registry is installed on controller nodes ( need to set up replication between them via rsync or something )
- `openstack-docker` role attempts to fetch container from registry, if it doesn't exist it will build it using giftwrap and upload.
- alternatives are used to create bash functions in `/etc/profile.d/` to do a docker run for the various commands we want to expose to the host cli.
- config, logs, cacert are volume mounted into container at runtime
- uses host networking to avoid port mapping shenanigans.

### CODE REVIEW

File: doc/docker.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Support for running openstack projects in docker.
+
+Ursula has experimental support for running the various openstack projects in docker.   We do not recommend ( yet ) running the harder stuff like `nova-compute` and `neutron` in docker,  however projects like `heat` are perfect for it and any examplesin this doc will demonstrate with `heat`.
```
Comment (j2sol) [inline doc/docker.md]: `examplesin` needs a space :)

File: doc/docker.md [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+To see this experimental support in action run `ursula --vagrant envs/example/docker site.yml`
+
+Ursula assumes you built your images with `giftwrap` or wish to.   It will run a docker registry for you ( on controller nodes ) if you set `docker_registry.enabled=True` , but does not currently replicate the registry data dir across both.
+
+To enabled heat to build and run via docker you will set `heat.openstack_install_method=docker`.   There are some docker related defaults in `roles/heat/defaults/main.yml` if you wish to override specific versions or not use the built in registry.
+
+the `openstack-docker` role does all the work prepare a docker image to be run for the given project.
```
Comment (j2sol) [inline doc/docker.md]: `all the work prepare` -> `all the work to prepare`

File: doc/docker.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+## role/openstack-docker
+
+First it will try and pull the docker image specified in the heat defaults.  If this does not exist then it will attempt to build an image for you with `giftwrap` based on hints provided from the `heat` defaults.
```
Comment (j2sol) [inline doc/docker.md]: This seems strange, talking about heat in a header for the openstack-docker role, and the header comes after explaining what the openstack-docker role does...

File: library/docker.py [hunk 4]
```diff
@@ -0,0 +1,1745 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+######################################################################
+
+DOCUMENTATION = '''
+---
+module: docker
+version_added: "1.4"
```
Comment (j2sol) [inline library/docker.py]: Is this coming from upstream ansible? Are we modifying it in some way?

File: library/docker_image.py [hunk 5]
```diff
@@ -0,0 +1,332 @@
...
+# You should have received a copy of the GNU General Public License
+# along with this software.  If not, see <http://www.gnu.org/licenses/>.
+
+######################################################################
+
+DOCUMENTATION = '''
+---
+module: docker_image
+author: "Pavel Antonov (@softzilla)"
+version_added: "1.5"
```
Comment (j2sol) [inline library/docker_image.py]: Same question about this. Why are we including it in ursula?

File: roles/apt-repos/defaults/main.yml [hunk 6]
```diff
@@ -2,8 +2,8 @@
 apt_repos:
   # All key_urls should be HTTPS to ensure we are retrieving the key we are expecting too
   docker:
-    repo: 'https://get.docker.com/ubuntu'
-    key_url: 'https://keyserver.ubuntu.com/pks/lookup?op=get&search=0xD8576A8BA88D21E9'
+    repo: 'https://apt.dockerproject.org/repo'
+    key_url: 'http://keyserver.ubuntu.com/pks/lookup?op=get&search=0x58118E89F3A912897C070ADBF76221572C52609D'
```
Comment (j2sol) [inline roles/apt-repos/defaults/main.yml]: Why https -> http?

File: roles/docker-registry/defaults/main.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
+---
+docker_registry:
+  type: opensource
+  ip: "{{ primary_ip }}"
+  port: 3001
+  secret: mydevsecret
+  storage:
+    type: filesystem
+    swift:
+      authurl: https://tardis.openstack.blueboxgrid.com:5001/v2.0
```
Comment (j2sol) [inline roles/docker-registry/defaults/main.yml]: Do we want this url sitting in here?

File: roles/docker-registry/handlers/main.yml [hunk 8]
```diff
@@ -0,0 +1,9 @@
+---
+- name: start docker registry
+  service: name=docker-registry state=started
+
+- name: stop docker registry
+  service: name=docker-registry state=stopped
+
+- name: restart docker registry
+  service: name=docker-registry state=restarted
```
Comment (j2sol) [inline roles/docker-registry/handlers/main.yml]: stopped and restarted handlers don't seem to be used, however restarted handler should maybe be used when configuration changes? Stopped could just go away.

File: roles/docker-registry/tasks/opensource.yml [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+  file: dest=/etc/docker/registry/opensource state=directory
+        recurse=true
+
+- name: create docker registry data path
+  file: dest={{ docker_registry.data_path }} state=directory
+        recurse=true
+
+- name: write out registry config files
+  template: src={{ item }} dest=/etc/docker/registry/opensource mode=0644
+  with_fileglob: ../templates/etc/docker/registry/opensource/*
```
Comment (j2sol) [inline roles/docker-registry/tasks/opensource.yml]: Should the service be restarted if these change?

File: roles/docker-registry/templates/etc/docker/registry/opensource/config.yml [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  filesystem:
+    rootdirectory: /var/lib/registry
+{% endif %}
+http:
+  addr: 0.0.0.0:5000
+  secret: {{ docker_registry.secret }}
+{% if docker_registry.tls.enabled|bool %}
+  tls:
+    certificate: /path/to/x509/public
+    key: /path/to/x509/private
```
Comment (j2sol) [inline roles/docker-registry/templates/etc/docker/registry/opensource/config.yml]: Shouldn't these use the variables from the role defaults?

File: roles/docker/tasks/main.yml [hunk 11]
```diff
@@ -1,8 +1,19 @@
...
+  pip: name=docker-py
+
+- name: configure docker service
+  template: src=etc/default/docker dest=/etc/default/docker
+  notify: restart docker service
+
+- name: make sure docker is running
+  service: name=docker state=started
+
+- meta: flush_handlers
```
Comment (j2sol) [inline roles/docker/tasks/main.yml]: The flush should come before the final `state=started` task.

File: library/docker_image.py (hunk 5 above)
Comment (paulczar) [inline library/docker_image.py]: this is modified to include the ability to pull an image

File: library/docker.py (hunk 4 above)
Comment (paulczar) [inline library/docker.py]: I don't think I ended up changing it,  can probably remove.

**Comment by paulczar** [comment]
closing all PRs with no recent activity.

**Pull request closed without merging**